src/groupeddataframe/groupeddataframe.jl

```diff
@@ -12,12 +12,22 @@ Not meant to be constructed directly, see `groupby`.
 """
 mutable struct GroupedDataFrame{T<:AbstractDataFrame}
     parent::T
-    cols::Vector{Int}                  # columns used for grouping
-    groups::Vector{Int}                # group indices for each row in 0:ngroups, 0 skipped
-    idx::Union{Vector{Int},Nothing}    # indexing vector sorting rows into groups
-    starts::Union{Vector{Int},Nothing} # starts of groups after permutation by idx
-    ends::Union{Vector{Int},Nothing}   # ends of groups after permutation by idx
-    ngroups::Int                       # number of groups
+    cols::Vector{Int}                    # columns used for grouping
+    groups::Vector{Int}                  # group indices for each row in 0:ngroups, 0 skipped
+    idx::Union{Vector{Int},Nothing}      # indexing vector sorting rows into groups
+    starts::Union{Vector{Int},Nothing}   # starts of groups after permutation by idx
+    ends::Union{Vector{Int},Nothing}     # ends of groups after permutation by idx
+    ngroups::Int                         # number of groups
+    keymap::Union{Dict{Any,Int},Nothing} # mapping of key tuples to group indices
+end
+
+function genkeymap(gd, cols)
+    d = Dict{Any,Int}()
+    sizehint!(d, length(gd.starts))
+    for (i, s) in enumerate(gd.starts)
```

Comment:
I guess this wouldn't hurt?
```suggestion
    @inbounds for (i, s) in enumerate(gd.starts)
```